### DISCUSSION TITLE
Setting date/recency boost, date weight in search / opensearch

### DISCUSSION THREAD

**Discussion by ssyberg** [wagtail/wagtail#12115] [General]
I'm working on elastic/open search integration and I'm trying to figure out how to weight dates (either published dates or perhaps a custom date field on a model). I don't see this mentioned in the Wagtail documentation but there is some mention in the elastic docs (https://www.elastic.co/guide/en/app-search/current/boosts.html#boosts-recency-boosts). Does anyone have any recommendations or plans to integrate into Wagtail?

**Comment by zerolab** [comment]
I believe https://docs.wagtail.org/en/stable/topics/search/indexing.html#index-searchfield should help? Pass the boost config via the `es_extra` kwarg

**Reply by ssyberg** [reply]
Thanks, yea I'm aware of `es_extra` but have no idea what goes there.

**Reply by ssyberg** [reply]
One thing that is confusing here is that we can set a `boost` value, but index boosting was deprecated back in Elastic Search 5.0, so I'm not sure if the `boost` value is being used at query time instead. If so I'd assume you could configure some `es_extra` boost configuration for the query time boost but when I try adding a `boost` attribute to the `es_extra` object I get this: `'mapper_parsing_exception', 'unknown parameter [boosts] on mapper`

**Reply by zerolab** [reply]
The boost kwarg on `SearchField` is now a query time boost (as of #10653)

`es_extra` is added in https://github.com/wagtail/wagtail/blob/0932ba395c9c27a5655f5f1eb6735e7d36c7aa8d/wagtail/search/backends/elasticsearch7.py#L194-L196 which is then added to the mappings in https://github.com/wagtail/wagtail/blob/0932ba395c9c27a5655f5f1eb6735e7d36c7aa8d/wagtail/search/backends/elasticsearch7.py#L209-L211

Thinking of it, you may needs a custom [`Elasticsearch7SearchQueryCompiler`](https://github.com/wagtail/wagtail/blob/0932ba395c9c27a5655f5f1eb6735e7d36c7aa8d/wagtail/search/backends/elasticsearch7.py#L456) that updates the end query to provide the additional boosts. To do so you'd need to define a custom backend, and then set it in your `WAGTAILSEARCH_BACKENDS` config under the `BACKEND` key

**Reply by ssyberg** [reply]
Ok if I'm understanding correctly, it seems like the `es_extra` object can set any attribute defined here: https://www.elastic.co/guide/en/elasticsearch/reference/7.8/mapping-params.html

**Reply by ssyberg** [reply]
> Thinking of it, you may needs a custom [Elasticsearch7SearchQueryCompiler](https://github.com/wagtail/wagtail/blob/0932ba395c9c27a5655f5f1eb6735e7d36c7aa8d/wagtail/search/backends/elasticsearch7.py#L456) that updates the end query to provide the additional boosts. To do so you'd need to define a custom backend, and then set it in your WAGTAILSEARCH_BACKENDS config under the BACKEND key

This sounds correct which I think is full circle back to my original question. Sounds like there is no way to inject any sort of boost functionality (beyond a simple boost numeric value) into Wagtail at the moment.

**Reply by zerolab** [reply]
heh, I suppose you're right. I was hoping `es_extra` would do it, but alas..

**Reply by ssyberg** [reply]
Well it's worthwhile to get confirmation I'm not missing something, thanks @zerolab !

**Comment by ssyberg** [comment]
This appears to be working for me, thanks for the idea @zerolab 
```python
from wagtail.search.backends.elasticsearch7 import (
    Elasticsearch7SearchQueryCompiler,
    Elasticsearch7SearchBackend,
)
from datetime import datetime


class CustomElasticsearch7SearchQueryCompiler(Elasticsearch7SearchQueryCompiler):
    def get_query(self):
        inner_query = self.get_inner_query()
        filters = self.get_filters()

        # Get the current date
        current_date = datetime.now().isoformat()

        # Apply the function_score with gauss decay function to boost recency
        function_score_query = {
            "function_score": {
                "query": {
                    "bool": {
                        "must": inner_query,
                        "filter": filters if filters else None,
                    }
                },
                "functions": [
                    {
                        "gauss": {
                            "first_published_at_filter": {
                                "origin": current_date,
                                "scale": "30d",
                                "decay": 0.9,
                            }
                        },
                        "weight": 1.2,
                    }
                ],
                "boost_mode": "sum",
            }
        }

        return function_score_query


class CustomElasticsearch7SearchBackend(Elasticsearch7SearchBackend):
    query_compiler_class = CustomElasticsearch7SearchQueryCompiler
```
